src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java

```diff
@@ -264,24 +263,27 @@ protected BindingCalculator(boolean forArguments) {
             this.storageCalculator = new StorageCalculator(forArguments);
         }
 
+        private static final int COPY_CHUNK_SIZE = 8;
+
         protected void spillStructUnbox(Binding.Builder bindings, MemoryLayout layout) {
             // If a struct has been assigned register or HFA class but
             // there are not enough free registers to hold the entire
             // struct, it must be passed on the stack. I.e. not split
             // between registers and stack.
 
             long offset = 0;
+            long align = layout.byteAlignment();
             while (offset < layout.byteSize()) {
-                long copy = Math.min(layout.byteSize() - offset, STACK_SLOT_SIZE);
-                VMStorage storage =
-                    storageCalculator.stackAlloc(copy, STACK_SLOT_SIZE);
-                if (offset + STACK_SLOT_SIZE < layout.byteSize()) {
+                long copy = Math.min(layout.byteSize() - offset, COPY_CHUNK_SIZE);
+                VMStorage storage = storageCalculator.stackAlloc(copy, align);
+                align = 1; // only first copy needs aligning
```

Comment:
so, this was an issue in the old code?

Reply:
Sorry, forgot to explain this.

In the old code we always passed STACK_SLOT_SIZE. This works out in the current implementation since we only allow values to be copied to the stack at stack slot aligned offset. Though, technically, we might need a more precise alignment in the case of Mac/AArch64. For instance, if a struct has 4 byte alignment, we might need to spill it at a 4-byte aligned stack offset.

I noticed this when looking through the code and decided to fix this. But, I'm think now that it kinda doesn't belong in this patch. I can remove it. All we need is the change to drop `SharedUtils.alignment` for `layout.byteAlignment` above.

Reply:
I think this might have slipped over from when I was working on the Mac/AArch64 stack spill issue, and I didn't think to remove it :/